Allow changing start/end times by dragging

fwitzke opened the pull request:
When selecting a single bar, it adds two handles to that bar so that the user can adjust start/end times.

fwitzke committed:
Allow dragging when selecting a single bar

fwitzke committed:
Extract Y computation function

fwitzke committed:
WIP: try to add/remove brush dynamically

fwitzke committed:
Add hacky workaround to avoid the side effect of removing the brush

fwitzke committed:
Disable dragging behaviour when clearing brush

fwitzke committed:
Prevent dragging startedAt > endedAt and endedAt < startedAt

fwitzke committed:
Avoid `startedAt` overlaps with other bars

fwitzke committed:
Avoid `endedAt` overlaps with other bars

fwitzke committed:
Add `onBarChanged` callback to allow notifying the client when dragging

fwitzke committed:
Remove unused function

fwitzke committed:
Handle bar click events to enable selecting a bar even without brush

mcramm commented on main.js:
```diff
@@ -81,7 +116,123 @@ var createChart = function createChart (element, data, opts) {
+        .attr('x', newX - (dragBarSize / 2));
+
+      opts.onBarChanged(d);
+    }
```
These `onDragRight` and `onDragLeft` functions are doing very similar things. The first three lines of each are (virtually) identical aside from what `d.<time>` they use to compute `newX`.

It might be time to start giving names to groups of operations by extracting them out into their own functions, even if they're only going to be used once. Something like:

```
function onDragRight (bar) {
      var newValue = getNewValue(bar); // Or heck, just create a whole new "bar"
      // var newBar = buildBar(bar, d3);

      // Maybe combine the two guards...
      if (newTime.x <= bar.startedAt) { return; }
      if (overlapWithFutureBar(bar, newValue)) { return; }

      bar.endedAt = newValue.time;

      // arguable whether these should be pulled out or not.
      d3.select('rect.selected')
        .attr('width', computeBarWidth);

      d3.select('rect#dragRight')
        .attr('x', newValue.x - (dragBarSize / 2));

      opts.onBarChanged(bar);
    }
```

It might also be time (in a different PR) to start pulling some behaviour out of `main.js`.

Commit:
Set SHELL in Makefile (gets `make lint` working)

Commit:
WIP - Abstract out Bar & overlap detection

Also adds in some tests for these.

Commit:
Remove unused `byLabel` function

Commit:
Merge branch 'test-drag-proposal' into drag-walks-into-a-bar

fwitzke committed:
Export a TimelineChart object

fwitzke committed:
Disable brush events before removing brush

There is an issue where removing the brush from inside the brush
callback ends up triggering another `brush` event. So when adding the
brush we attach the listeners, and remove them when removing the brush
